storage: rework load-based splitting

tbg opened the pull request:
The old code was difficult to reason about and interspersed with the
replica code. Factor out a component Decider which encapsulates the
state and logic of load-based splitting.

The reason I originally looked at this was that I had introduced a guard
in the split queue that was supposed to check whether the load based
split was still recommended (time may have passed since it was
originally computed). This was frail since the method I called would
only recommend the split once, even though this was only obvious upon
closer examination. The new API prevents such mistakes.

Fixes #34114.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/34221)
<!-- Reviewable:end -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, just the stale comment

> Reviewed 10 of 10 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34221#-:-LX4j1I3BnoGncexrF-s:bg6d9iy)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/storage/replica.go, line 776 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34221#-LX4dL9W13ujJTN4n5cW:-LX4dL9W13ujJTN4n5cX:b-lbzikx) ([raw file](https://github.com/cockroachdb/cockroach/blob/70fe42c570bf8c58e795f0ca52ea1965ca4dcd26/pkg/storage/replica.go#L776)):*
> ```Go
> // GetSplitQPS returns the Replica's queries/s request rate.
> // The value returned represents the QPS recorded at the time of the
> // last request which can be found using GetLastRequestTime().
> ```

Stale comment. GetLastRequestTime no longer exists.


<!-- Sent from Reviewable.io -->

tbg committed:
storage: rework load-based splitting

The old code was difficult to reason about and interspersed with the
replica code. Factor out a component Decider which encapsulates the
state and logic of load-based splitting.

The reason I originally looked at this was that I had introduced a guard
in the split queue that was supposed to check whether the load based
split was still recommended (time may have passed since it was
originally computed). This was frail since the method I called would
only recommend the split once, even though this was only obvious upon
closer examination. The new API prevents such mistakes.

Fixes #34114.

Release note: None
